### DISCUSSION TITLE
Compiler plugin - deferring failure to a separate task

### DISCUSSION THREAD

**Discussion by matejdro** [detekt/detekt#6084] [General]
I've just heard about the upcoming compiler plugin and from what I understand, the plugin will analyze and fail the compilation completely if detekt issues are detected. I see two problems with that:

* WIP in-development code is often messy and it would slow down development considerably, if it had to be detekt-compliant all the time. Cleaning up the code before commiting and letting detekt analyze in pre-commit hook seems a better tradeoff.
* If build fails because of detekt error, all dependant modules will not be build, meaning that detekt analysis will not run on dependant modules, so developer cannot easily get a list of all issues in the project

What if compiler plugin would just write report files, but not actually fail the build? Then something like `analyzeDetektReport` task could be made that would read the reports and fail with the console output (similar to how `detekt` task fails now).

**Comment by BraisGabin** [comment]
> WIP in-development code is often messy

I completely agree with this. I don't want all my compilation to be detekt-ready. Even on detekt we don't want them to be warning-free. For that reason we only enable this flag on CI instead of having it enable by default: https://github.com/detekt/detekt/blob/e41f9f23d5b5c86f86f41d512563eea8931665e3/.github/workflows/pre-merge.yaml#L103 

I can't talk to loud here because I didn't test the compiler plugin yet and I'm not 100% sure how it works. But in my daily builds I don't want to have the detekt plugin enabled. I want really fast compilations. The compiler plugin makes detekt way faster. But don't run detekt is even faster. So it would be nice to be able to have the plugin disable by default and only enable it when you really want detekt to pass (pre-commit, pre-push, pre-merge...). As I said, maybe all of this is doable right now, I have no idea.

About when to raise the errors: On detekt 1.x I think it will be really difficult to archive that. I raised this proposal for detekt 2.0: https://github.com/detekt/detekt/discussions/6085. It should allow to do what you are asking for.

**Reply by matejdro** [reply]
I also admit that I have not tried the compiler plugin yet, my knowledge of this mostly stems from the Kotlinconf talk.

I think problem with enabling detekt compiler plugin only during pre-commit tasks is that kotlin compiler will not actually run most of the time (code would be UP-TO-DATE since most of the time you don't actually change anything between build and commit) and thus detekt would also not run.

**Reply by BraisGabin** [reply]
I would asume that gradle is smart enoght to know that a compilation with a plugin could lead to different output than without. I mean, the enabled plugins should be an input for that task. If that's not the case we should raise an issue somewhere because that would not be our fault.

**Reply by matejdro** [reply]
Ah, you are right.
